common/transport/amqp/src/amqp.ts

```diff
@@ -145,44 +235,190 @@ export class Amqp {
-                this._amqp.removeListener(_amqpClientError, connectErrorHandler);
-                /*Codes_SRS_NODE_COMMON_AMQP_16_003: [The `connect` method shall call the `done` callback if the connection fails.] */
-                this._fsm.transition('disconnected', connectCallback, connectError || err);
-              });
+          _onEnter: (connectionParameters, connectCallback) => {
+            this._rheaContainer.options.sender_options = {properties: {'com.microsoft:client-version': this._config.userAgentString}, reconnect: false};
+            this._rheaContainer.options.receiver_options = {properties: {'com.microsoft:client-version': this._config.userAgentString}, reconnect: false, autoaccept: autoSettleMode};
+            this._connectionCallback = connectCallback;
+            this._indicatedConnectionError = undefined;
+            this._disconnectionOccurred = false;
+            this._sessionCloseOccurred = false;
+            this._connectionCloseOccurred = false;
+            //
+            // According to the rhea maintainers, one can depend on that fact that no actual network activity
+            // will occur until the nextTick() after the call to connect.  Because of that, one can
+            // put the event handlers on the rhea connection object returned from the connect call and be assured
+            // that the listeners are in place BEFORE any possible events will be emitted on the connection.
+            //
+            this._rheaConnection = this._rheaContainer.connect(connectionParameters);
+            manageConnectionHandlers('on');
+          },
+          'connection_open': (context: EventContext) => {
+            debug('in the connection open handler for the connection state: ' + context.connection.name);
+            this._rheaConnection = context.connection;
+            let callback = this._connectionCallback;
+            this._connectionCallback = undefined;
+            this._fsm.transition('connecting_session', callback);
+          },
+          'connection_close': (context: EventContext) => {
+            debug('In the connection_close handler for the connecting state: ' + context.connection.name + ' with an indicated error of: ' + this._getErrorName(this._indicatedConnectionError));
+            let err = this._indicatedConnectionError;
+            let callback = this._connectionCallback;
+            this._indicatedConnectionError = undefined;
+            this._connectionCallback = undefined;
+            this._connectionCloseOccurred = true;
+            manageConnectionHandlers('removeListener');
+            this._fsm.transition('disconnected', callback, err);
+          },
+          'connection_error': (context: EventContext) => {
+            debug('In the connection_error handler for the connecting state: ' + context.connection.name + ' with error of: ' + this._getErrorName(context.connection.error));
+            debug('Deferring transition until the connection_close occurs.');
```

Comment:
>debug('Deferring transition until the connection_close occurs.'); [](start = 12, length = 65)

you're not deferring anything here, i think the debug statement is misleading especially considering how "defer" means something else in the context of machina.js

Reply:
It is stating exactly what I meant.  Will remove these debugs.

---
In reply to: [209132217](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209132217) [](ancestors = 209132217)